### PR TITLE
spu/arm64: clean up assembly code generation

### PR DESCRIPTION
**Pull request by sguo35**
Clean up asmjit usage so we don't unnecessarily allocate/leak memory anymore for SPURecompiler functions.

Tested with spurs test and 2 basic games and doesn't appear to cause any regressions.

### CODE REVIEW

**Comment by elad335** [comment]
Hi, you mention a "leak" but the affected code is called once during program initialization. (they are static variables)
We are talking about just a few dozen bytes here which is very far away from a "leak".

**Comment by sguo35** [comment]
> Hi, you mention a "leak" but the affected code is called once during program initialization. (they are static variables) We are talking about just a few dozen bytes here which is very far away from a "leak".

Neko [told me to fix it ](https://github.com/RPCS3/rpcs3/pull/12338#issuecomment-1185403438) after the SPU arm port was landed. 

Also, some of this code is called every time we branch to a new function. We can run out of the 256MB allocated JIT memory in some games with large SPU code size as a result of the memory bloat (some Arm code here is a lot bigger than x86 versions in code size).

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 1]
```diff
@@ -1636,18 +1737,24 @@ void spu_recompiler_base::dispatch(spu_thread& spu, void*, u8* rip)
 
 		atomic_storage<u64>::release(*reinterpret_cast<u64*>(rip - 8), result);
 #elif defined(ARCH_ARM64)
-		auto jump_instrs = build_function_asm<spu_function_t>("", [](native_asm& c, auto& args)
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:1639]: Oh I see, here is the leak. I think this is the only place where it should be changed. (bytecode is much less readable and is more difficult to work with)

Comment (sguo35) [reply rpcs3/Emu/Cell/SPURecompiler.cpp:1639]: I see. I've removed the other changes.

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 2]
```diff
@@ -1636,18 +1636,24 @@ void spu_recompiler_base::dispatch(spu_thread& spu, void*, u8* rip)
...
-				c.bind(branch_target);
-				c.embedUInt64(reinterpret_cast<u64>(spu_runtime::tr_all));
-			});
-		u128 result = *reinterpret_cast<u128*>(jump_instrs);
+		union
+		{
+			u8 bytes[16];
+			u128 result;
+		};
+		// ldr x9, #8
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:1644]: Newline above.

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 3]
```diff
@@ -1636,18 +1636,24 @@ void spu_recompiler_base::dispatch(spu_thread& spu, void*, u8* rip)
...
+		bytes[0] = 0x49;
+		bytes[1] = 0x00;
+		bytes[2] = 0x00;
+		bytes[3] = 0x58;
+		// br x9
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:1649]: Newline above.

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 4]
```diff
@@ -1768,18 +1774,24 @@ void spu_recompiler_base::branch(spu_thread& spu, void*, u8* rip)
...
-			c.bind(branch_target);
-			c.embedUInt64(reinterpret_cast<u64>(func));
-		});
-	u128 result = *reinterpret_cast<u128*>(jmp_instrs);
+	union
+	{
+		u8 bytes[16];
+		u128 result;
+	};
+	// ldr x9, #8
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:1782]: Newline above.

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 5]
```diff
@@ -1768,18 +1774,24 @@ void spu_recompiler_base::branch(spu_thread& spu, void*, u8* rip)
...
+	bytes[0] = 0x49;
+	bytes[1] = 0x00;
+	bytes[2] = 0x00;
+	bytes[3] = 0x58;
+	// br x9
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:1787]: Newline above.

File: rpcs3/Emu/Cell/SPURecompiler.cpp (hunk 1 above)
Comment (elad335) [reply rpcs3/Emu/Cell/SPURecompiler.cpp:1639]: And keep the rebuild_ubertrampoline changes as well.

**Comment by sguo35** [comment]
I also readded the `make_branch_patchpoint` changes as that gets called quite often too.

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 6]
```diff
@@ -1038,43 +1038,70 @@ spu_function_t spu_runtime::rebuild_ubertrampoline(u32 id_inst)
...
+				*raw++ = 0x20;
+				*raw++ = 0x03;
+				*raw++ = 0xD5;
+
+				// nop
+				*raw++ = 0x1F;
+				*raw++ = 0x20;
+				*raw++ = 0x03;
+				*raw++ = 0xD5;
+			} else
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:1070]: Else on a new line.

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 7]
```diff
@@ -1038,43 +1038,70 @@ spu_function_t spu_runtime::rebuild_ubertrampoline(u32 id_inst)
...
+				case asmjit::arm::CondCode::kUnsignedLT:
+					*raw++ = 0x43;
+					break;
+				case asmjit::arm::CondCode::kUnsignedGT:
+					*raw++ = 0x48;
+					break;
+				default:
+					asm("brk 0x42");
+				}
+				*raw++ = 0x00;
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:1084]: Newline above.

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 8]
```diff
@@ -1573,34 +1606,60 @@ spu_function_t spu_runtime::make_branch_patchpoint(u16 data) const
...
 
-			c.adr(a64::x21, replace_addr);
-			// 16 byte alignment for the jump replacement
-			c.nop();
-			c.nop();
-			c.nop();
-			Label branch_target = c.newLabel();
+	if (!raw)
+	{
+		return nullptr;
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:1618]: You can replace with `u8* const patch_fn = ensure(jit_runtime::alloc(36, 16));` to assert it being non-nullptr.

File: rpcs3/Emu/Cell/SPURecompiler.cpp [hunk 9]
```diff
@@ -1038,43 +1038,70 @@ spu_function_t spu_runtime::rebuild_ubertrampoline(u32 id_inst)
...
-						c.bind(cont);
-					}
-				});
-			u8 mem_used = 7 * 4;
-			memcpy(raw, reinterpret_cast<u8*>(temp), mem_used);
-			raw += mem_used;
+			// taddr
+			std::memcpy(raw, &taddr, 8);
+			raw += 8;
+			// cont
```
Comment (elad335) [inline rpcs3/Emu/Cell/SPURecompiler.cpp:1104]: Newline above, also naybe mention that this is supposed to be a code label.

**Pull request merged**